File: frontend/src/lib/validations/member.ts
import { z } from "zod";

/**
 * Normalize an Indian phone number to canonical 10-digit format.
 *
 * Mirrors the backend's `normalize_phone()` so the user sees the same
 * validation result in the browser before the form is submitted.
 * Strips: +91 prefix, leading 0, spaces, dashes, parentheses.
 */
export function normalizePhone(raw: string): string {
  // Remove all non-digit characters
  let digits = raw.replace(/\D/g, "");
  // Strip country code prefix (91) if present and result would be 10 digits
  if (digits.length === 12 && digits.startsWith("91")) {
    digits = digits.slice(2);
  }
  // Strip leading 0 (some users type 09876...)
  if (digits.length === 11 && digits.startsWith("0")) {
    digits = digits.slice(1);
  }
  return digits;
}

/**
 * Zod schema for member form validation.
 * Mirrors backend Pydantic MemberCreateRequest exactly.
 *
 * Phone pattern: Indian mobile (starts with 6-9, 10 digits)
 * Phone is normalized before validation (strips +91, spaces, dashes).
 */
export const memberFormSchema = z.object({
  name: z
    .string()
    .min(2, "Name must be at least 2 characters")
    .max(200, "Name is too long"),
  phone: z
    .string()
    .transform(normalizePhone)
    .pipe(z.string().regex(/^[6-9]\d{9}$/, "Enter a valid 10-digit Indian mobile number")),
  email: z.union([z.literal(""), z.string().email("Enter a valid email")]).optional(),
  gender: z.enum(["male", "female", "other", ""]).optional(),
  father_name: z.string().max(200).optional().or(z.literal("")),
  batch: z.enum(["morning", "evening", "afternoon", ""]).optional(),
  membership_plan: z.string().max(100).optional().or(z.literal("")),
  membership_start: z.string().optional().or(z.literal("")),
  membership_end: z.string().optional().or(z.literal("")),
  amount_paid: z.number().min(0, "Amount cannot be negative").optional().default(0),
});

export type MemberFormValues = z.infer<typeof memberFormSchema>;

File: backend/app/schemas/member.py
from datetime import date, datetime
from uuid import UUID
import re

from pydantic import BaseModel, EmailStr, Field, field_validator

from app.models.member import Batch, Gender, MembershipStatus
from app.schemas.sanitize import strip_html_tags


def normalize_phone(raw: str) -> str:
    """Normalize Indian phone numbers to canonical 10-digit format.

    Accepts common user input variations:
      +91 9876543210 → 9876543210
      091-9876543210 → 9876543210
      (0)9876543210  → 9876543210

    Strips +91 prefix, leading 0, spaces, dashes, and parentheses.
    Validation (must start with 6-9) is done after normalization.
    """
    digits = re.sub(r"\D", "", raw)
    if len(digits) == 12 and digits.startswith("91"):
        digits = digits[2:]
    if len(digits) == 11 and digits.startswith("0"):
        digits = digits[1:]
    return digits


class MemberCreateRequest(BaseModel):
    name: str = Field(..., min_length=2, max_length=200)
    phone: str = Field(..., min_length=1)
    email: EmailStr | None = None
    gender: Gender | None = None
    date_of_birth: date | None = None
    father_name: str | None = Field(None, max_length=200)
    batch: Batch | None = None
    emergency_contact: str | None = None
    membership_plan: str | None = None
    membership_start: date | None = None
    membership_end: date | None = None
    amount_paid: int = 0  # in paise

    @field_validator("phone")
    @classmethod
    def validate_and_normalize_phone(cls, v: str) -> str:
        normalized = normalize_phone(v)
        if not re.match(r"^[6-9]\d{9}$", normalized):
            raise ValueError("Enter a valid 10-digit Indian mobile number")
        return normalized

    @field_validator("name", "father_name", "emergency_contact", "membership_plan")
    @classmethod
    def sanitize_text_fields(cls, v: str | None) -> str | None:
        return strip_html_tags(v) if v else v


class MemberUpdateRequest(BaseModel):
    name: str | None = Field(None, min_length=2, max_length=200)
    phone: str | None = Field(None, min_length=1)
    email: EmailStr | None = None
    gender: Gender | None = None
    father_name: str | None = Field(None, max_length=200)
    batch: Batch | None = None
    # membership_status is NOT updatable directly — use membership lifecycle APIs
    membership_plan: str | None = None
    membership_start: date | None = None
    membership_end: date | None = None
    amount_paid: int | None = None
    # Optimistic locking: client sends the version it last read.
    # Server rejects the update with 409 Conflict if another edit landed first.
    # Optional (None) for backward compatibility with older clients.
    version: int | None = None

    @field_validator("phone")
    @classmethod
    def validate_and_normalize_phone(cls, v: str | None) -> str | None:
        if v is None:
            return v
        normalized = normalize_phone(v)
        if not re.match(r"^[6-9]\d{9}$", normalized):
            raise ValueError("Enter a valid 10-digit Indian mobile number")
        return normalized

    @field_validator("name", "father_name", "membership_plan")
    @classmethod
    def sanitize_text_fields(cls, v: str | None) -> str | None:
        return strip_html_tags(v) if v else v


class MemberResponse(BaseModel):
    id: UUID
    name: str
    phone: str
    email: str | None
    gender: Gender | None
    father_name: str | None = None
    batch: Batch | None = None
    membership_status: MembershipStatus
    membership_plan: str | None
    membership_start: date | None
    membership_end: date | None
    amount_paid: int
    photo_url: str | None = None
    version: int = 0
    created_at: datetime | None = None  # Included for audit trail visibility in UI
    updated_at: datetime | None = None  # Included for audit trail visibility in UI

    model_config = {"from_attributes": True}


class MemberListResponse(BaseModel):
    members: list[MemberResponse]
    total: int
